test/MathOptInterface/callback.jl

```diff
@@ -0,0 +1,548 @@
+# Copyright (c) 2023: COPT-Public
+#
+# Use of this source code is governed by an MIT-style license that can be found
+# in the LICENSE.md file or at https://opensource.org/licenses/MIT.
+
+module TestCallback
+
+using COPT
+using MathOptInterface
+using Random
+using Test
+
+const MOI = MathOptInterface
+
+function runtests()
+    for name in names(@__MODULE__; all = true)
+        if startswith("$(name)", "test_")
+            @testset "$(name)" begin
+                getfield(@__MODULE__, name)()
+            end
+        end
+    end
+    return
+end
+
+function callback_simple_model()
+    env = COPT.Env()
+    model = COPT.Optimizer(env)
+    MOI.set(model, MOI.RawOptimizerAttribute("LogToConsole"), 0)
+    MOI.Utilities.loadfromstring!(
+        model,
+        """
+    variables: x, y
+    maxobjective: y
+    c1: x in Integer()
+    c2: y in Integer()
+    c3: x in Interval(0.0, 2.5)
+    c4: y in Interval(0.0, 2.5)
+""",
+    )
+    x = MOI.get(model, MOI.VariableIndex, "x")
+    y = MOI.get(model, MOI.VariableIndex, "y")
+    return model, x, y
+end
+
+function callback_knapsack_model()
+    env = COPT.Env()
+    model = COPT.Optimizer(env)
+    MOI.set(model, MOI.RawOptimizerAttribute("LogToConsole"), 0)
+    N = 30
+    x = MOI.add_variables(model, N)
+    MOI.add_constraints(model, x, MOI.ZeroOne())
+    MOI.set.(model, MOI.VariablePrimalStart(), x, 0.0)
+    Random.seed!(1)
+    item_weights, item_values = rand(N), rand(N)
+    MOI.add_constraint(
+        model,
+        MOI.ScalarAffineFunction(MOI.ScalarAffineTerm.(item_weights, x), 0.0),
+        MOI.LessThan(10.0),
+    )
+    MOI.set(
+        model,
+        MOI.ObjectiveFunction{MOI.ScalarAffineFunction{Float64}}(),
+        MOI.ScalarAffineFunction(MOI.ScalarAffineTerm.(item_values, x), 0.0),
+    )
+    MOI.set(model, MOI.ObjectiveSense(), MOI.MAX_SENSE)
+    return model, x, item_weights
+end
+
+# function test_lazy_constraint_callback()
```

Comment:
Intentionally commented? Are these the ones that error randomly?

Reply:
That particular test does not seem to crash with Julia 1.6.7, but the test below crashes in 1.6.7.  It does not crash with  Julia 1.9.0.
https://github.com/COPT-Public/COPT.jl/blob/cec3b34941249132f79235539237996ddc2a2991/test/MathOptInterface/callback.jl#L119

Reply:
Does COPT call the callback from multiple threads, or only from the main thread? The Julia/C interface is not thread safe. 

Do you have a log of the crash?

Reply:
This would also explain our observation, that running COPT with just 1 thread seems to work fine.

COPT calls the callbacks from multiple threads. However, the callback should never be called from 2 threads at the same time. We have a mutex in place that should prevent this. We experienced multiple phenomena here.
With 1.9 when adding prints to the callbacks via `println` we often ran into deadlocks where some println would be stuck deep down waiting for something. However, without prints we were successful in running the tests.

With 1.6.7 we ran into segfaults. I can add a partial bt from GDB (with julia compiled in debug):
```
Thread 10 "julia" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fff87154700 (LWP 5228)]
jl_excstack_state () at /home/nckempke/git/julia/src/rtutils.c:287
287	    jl_excstack_t *s = ptls->current_task->excstack;
(gdb) bt
#0  jl_excstack_state () at /home/nckempke/git/julia/src/rtutils.c:287
#1  0x00007fff9726e0cf in julia__copt_callback_wrapper_48 (p_model=93825040736496, cb_data=93824994782048, cb_context=2, p_user_data=140735617844720)
    at /home/nckempke/callbacks/COPT.jl/src/MOI/callback.jl:43
#2  0x00007fff9727a5a9 in jlcapi.copt_callback_wrapper_51 ()
#3 ... COPT
...

(gdb) f 1
#1  0x00007fff9726e0cf in julia__copt_callback_wrapper_48 (p_model=93825040736496, cb_data=93824994782048, cb_context=2, p_user_data=140735617844720)
    at /home/nckempke/callbacks/COPT.jl/src/MOI/callback.jl:43
43	    try
(gdb) f 0
#0  jl_excstack_state () at /home/nckempke/git/julia/src/rtutils.c:287
287	    jl_excstack_t *s = ptls->current_task->excstack;
(gdb) p ptls
$1 = (jl_ptls_t) 0x7fff8714c700
(gdb) p ptls->current_task 
$2 = (struct _jl_task_t *) 0x0
```
Frame 3 and everything above is on the COPT side and calls the julia callback. I printed the variable that is causing the segfault as well.

I can also try to get a stacktrace from the 'getting stuck' in Julia 1.9 in debug if you'd like.

Reply:
We explicitly set the number of threads to 1 in CPLEX when using callbacks:

https://github.com/jump-dev/CPLEX.jl/blob/ccde44778f93d65af3c6492e44571af03bc015cc/src/MOI/MOI_callbacks.jl#L101-L108
https://github.com/jump-dev/CPLEX.jl/blob/ccde44778f93d65af3c6492e44571af03bc015cc/src/MOI/MOI_callbacks.jl#L180-L186

In Julia versions <v1.9, the lack of thread safety was a fundamental limitation of C functions calling back into Julia. It was supposedly fixed in Julia v1.9:
https://github.com/JuliaLang/julia/blob/master/HISTORY.md#multi-threading-changes
https://github.com/JuliaLang/julia/pull/46609
But there are probably still some rough edges.

> With 1.9 when adding prints to the callbacks via println we often ran into deadlocks where some println would be stuck deep down waiting for something.

Yeah. My guess is that now you'll need to add a bunch of locks to make the Julia side thread safe. Printing accesses the global variable `stdout` with some buffers, so I could see how that'd go wrong.

I think Gurobi has things setup so that the callbacks are only ever executed from a single thread on the C side, even if there are multiple threads working in parallel behind the scenes.

For now, I'd just follow CPLEX and fix the threads to 1, then wait to see if anyone complains.

Reply:
> However, the callback should never be called from 2 threads at the same time. We have a mutex in place that should prevent this.

I guess that's true from the C side, but Julia might still async some things so its still, for example, buffering to stdout even after the Julia callback has returned to C.

Reply:
What happens if you add `flush(stdout)` after the `println` statements?

Reply:
Mh - I don't think adding a flush helped - I tried that but cannot quite remember the result. I think the thread got stuck already in the println, so never reached the flush. I'll try and reproduce the deadlock with 2 threads in COPT and print the trace of the stuck one.

Ok, so for Julia version < 1.9 we should set the COPT `Threads` to 1 in the interface then.

Reply:
Here is the backtrace of where the deadlock happens - this is a slightly different file than the one posted here (I attached it) - and we're only running COPT with 2 threads to narrow this down a bit. Thread 1 and 10 are 2 COPT threads. Thread 10 has been created in thread 1 by COPT. 1 is currently waiting for 10 to get to some synchronization point inside COPT. 10 however seems to be stuck in some Julia routine waiting for 1 possibly? It also seems like 10 has just been created, right before the `println` call.
```
...
*******************************
*** This is an opt version  ***
*******************************
Cardinal Optimizer v6.5.3, git hash: 6d632bfc0. Build date May 14 2023
Copyright Cardinal Operations 2023. All Rights Reserved
Setting parameter 'LogToConsole' to 1
Calling COPT_GetCallbackInfo in solution context
Called COPT_GetCallbackInfo in solution context
Calling COPT_GetCallbackInfo in solution context
Called COPT_GetCallbackInfo in solution context
Calling COPT_GetCallbackInfo in solution context
Called COPT_GetCallbackInfo in solution context
Calling COPT_GetCallbackInfo in solution context
Called COPT_GetCallbackInfo in solution context
[New Thread 0x7fff89140700 (LWP 43186)]
Calling COPT_GetCallbackInfo in solution context^C   <---- SIGINT to see where we deadlocked
Thread 2 "julia" received signal SIGINT, Interrupt.
[Switching to Thread 0x7ffff185a700 (LWP 43159)]
__pthread_kill (threadid=<optimized out>, signo=2) at ../sysdeps/unix/sysv/linux/pthread_kill.c:56
56	../sysdeps/unix/sysv/linux/pthread_kill.c: No such file or directory.
(gdb) bt
#0  __pthread_kill (threadid=<optimized out>, signo=2) at ../sysdeps/unix/sysv/linux/pthread_kill.c:56
#1  0x00007ffff720032d in jl_ignore_sigint () at /home/nckempke/git/julia/src/signal-handling.c:238
#2  0x00007ffff72018fc in signal_listener (arg=0x0) at /home/nckempke/git/julia/src/signals-unix.c:814
#3  0x00007ffff7f8d609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#4  0x00007ffff7eb2133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
(gdb) info threads
  Id   Target Id                                 Frame 
  1    Thread 0x7ffff7d70240 (LWP 43157) "julia" futex_abstimed_wait_cancelable (private=0, abstime=0x0, clockid=0, expected=0, futex_word=0x555555a486c8) at ../sysdeps/nptl/futex-internal.h:320
* 2    Thread 0x7ffff185a700 (LWP 43159) "julia" __pthread_kill (threadid=<optimized out>, signo=2) at ../sysdeps/unix/sysv/linux/pthread_kill.c:56
  3    Thread 0x7fffdeb42700 (LWP 43160) "julia" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7fffe0d259e0 <thread_status+96>) at ../sysdeps/nptl/futex-internal.h:183
  4    Thread 0x7fffd6341700 (LWP 43161) "julia" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7fffe0d25a60 <thread_status+224>) at ../sysdeps/nptl/futex-internal.h:183
  5    Thread 0x7fffcdb40700 (LWP 43162) "julia" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7fffe0d25ae0 <thread_status+352>) at ../sysdeps/nptl/futex-internal.h:183
  6    Thread 0x7fffcd33f700 (LWP 43163) "julia" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7fffe0d25b60 <thread_status+480>) at ../sysdeps/nptl/futex-internal.h:183
  7    Thread 0x7fffbcb3e700 (LWP 43164) "julia" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7fffe0d25be0 <thread_status+608>) at ../sysdeps/nptl/futex-internal.h:183
  8    Thread 0x7fffb433d700 (LWP 43165) "julia" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7fffe0d25c60 <thread_status+736>) at ../sysdeps/nptl/futex-internal.h:183
  9    Thread 0x7fffabb3c700 (LWP 43166) "julia" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7fffe0d25ce0 <thread_status+864>) at ../sysdeps/nptl/futex-internal.h:183
  10   Thread 0x7fff89140700 (LWP 43186) "julia" futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7fff84005810) at ../sysdeps/nptl/futex-internal.h:183
(gdb) thread 1
[Switching to thread 1 (Thread 0x7ffff7d70240 (LWP 43157))]
#0  futex_abstimed_wait_cancelable (private=0, abstime=0x0, clockid=0, expected=0, futex_word=0x555555a486c8) at ../sysdeps/nptl/futex-internal.h:320
320	../sysdeps/nptl/futex-internal.h: No such file or directory.
(gdb) bt
#0  futex_abstimed_wait_cancelable (private=0, abstime=0x0, clockid=0, expected=0, futex_word=0x555555a486c8) at ../sysdeps/nptl/futex-internal.h:320
#1  do_futex_wait (sem=sem@entry=0x555555a486c8, abstime=0x0, clockid=0) at sem_waitcommon.c:112
#2  0x00007ffff7f97548 in __new_sem_wait_slow (sem=0x555555a486c8, abstime=0x0, clockid=0) at sem_waitcommon.c:184
#3  0x00007fff89acc64a in ss_Signal_Wait (signal=0x555555a486c8) at tool_thread.c:131
...
COPT frames
...
#20 julia__optimize!_157 (model=...) at /home/nckempke/callbacks/COPT.jl/src/MOI/MOI_wrapper.jl:2846
#21 0x00007fffa1737844 in julia_optimize!_148 (model=...) at /home/nckempke/callbacks/COPT.jl/src/MOI/MOI_wrapper.jl:2890
...
more frames
(gdb) thread 10
[Switching to thread 10 (Thread 0x7fff89140700 (LWP 43186))]
#0  futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7fff84005810) at ../sysdeps/nptl/futex-internal.h:183
183	../sysdeps/nptl/futex-internal.h: No such file or directory.
(gdb) bt
#0  futex_wait_cancelable (private=<optimized out>, expected=0, futex_word=0x7fff84005810) at ../sysdeps/nptl/futex-internal.h:183
#1  __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x7fff840057c0, cond=0x7fff840057e8) at pthread_cond_wait.c:508
#2  __pthread_cond_wait (cond=0x7fff840057e8, mutex=0x7fff840057c0) at pthread_cond_wait.c:647
#3  0x00007ffff7290938 in uv_cond_wait (cond=0x7fff840057e8, mutex=0x7fff840057c0) at src/unix/thread.c:883
#4  0x00007ffff71e2261 in ijl_task_get_next (trypoptask=0x7fffe3a73db0 <jl_system_image_data+7600880>, q=0x7fff8c6cc050, checkempty=0x7fffe55c38c0 <jl_system_image_data+36238848>)
    at /home/nckempke/git/julia/src/partr.c:421
#5  0x00007fffe2b6f378 in julia_poptask_33132 (W=...) at task.jl:974
#6  0x00007fffe2b6f22b in julia_wait_44615 () at task.jl:983
#7  0x00007fffe3130fc1 in julia_uv_write_50041 (s=..., p=<optimized out>, n=<optimized out>) at stream.jl:1048
#8  0x00007fffe31317c2 in julia_unsafe_write_49033 (s=..., p=<optimized out>, n=<optimized out>) at stream.jl:1120
#9  0x00007fffe28b2b79 in write () at strings/io.jl:244
#10 print () at strings/io.jl:246
#11 japi1_print_41611 (io=..., xs...=<optimized out>) at strings/io.jl:46
#12 0x00007ffff717c685 in jl_fptr_args (f=0x7fffe6d07690 <jl_system_image_data+60634064>, args=0x7fff8913f0f8, nargs=3, m=0x7fffe3626e80 <jl_system_image_data+3091392>)
    at /home/nckempke/git/julia/src/gf.c:2405
#13 0x00007ffff717d994 in _jl_invoke (F=0x7fffe6d07690 <jl_system_image_data+60634064>, args=0x7fff8913f0f8, nargs=3, mfunc=0x7fffe3626e30 <jl_system_image_data+3091312>, world=33494)
    at /home/nckempke/git/julia/src/gf.c:2739
#14 0x00007ffff717e439 in ijl_apply_generic (F=0x7fffe6d07690 <jl_system_image_data+60634064>, args=0x7fff8913f0f8, nargs=3) at /home/nckempke/git/julia/src/gf.c:2940
#15 0x00007ffff718dd0f in jl_apply (args=0x7fff8913f0f0, nargs=4) at /home/nckempke/git/julia/src/julia.h:1879
#16 0x00007ffff7191656 in do_apply (args=0x7fff8913f2f0, nargs=4, iterate=0x7fffe6c7f880 <jl_system_image_data+60077504>) at /home/nckempke/git/julia/src/builtins.c:730
#17 0x00007ffff71916e4 in jl_f__apply_iterate (F=0x0, args=0x7fff8913f2e8, nargs=5) at /home/nckempke/git/julia/src/builtins.c:738
#18 0x00007fffe2765e6a in japi1_println_34017 (io=..., xs...=0x7fff8c6d0010) at strings/io.jl:75
#19 0x00007ffff717c685 in jl_fptr_args (f=0x7fffe3627170 <jl_system_image_data+3092144>, args=0x7fff8913f4b0, nargs=2, m=0x7fffe3627500 <jl_system_image_data+3093056>)
    at /home/nckempke/git/julia/src/gf.c:2405
#20 0x00007ffff717d994 in _jl_invoke (F=0x7fffe3627170 <jl_system_image_data+3092144>, args=0x7fff8913f4b0, nargs=2, mfunc=0x7fffe36274b0 <jl_system_image_data+3092976>, world=33494)
    at /home/nckempke/git/julia/src/gf.c:2739
#21 0x00007ffff717e439 in ijl_apply_generic (F=0x7fffe3627170 <jl_system_image_data+3092144>, args=0x7fff8913f4b0, nargs=2) at /home/nckempke/git/julia/src/gf.c:2940
#22 0x00007fffe2c92801 in julia_println_34004 (xs...=...) at coreio.jl:4
#23 0x00007fffa17447c2 in julia_load_callback_variable_primal_409 (cb_data=..., cb_context=2) at /home/nckempke/callbacks/COPT.jl/src/MOI/callback.jl:118
#24 0x00007fffa1744b2d in julia_callback_function_396 (cb_data=..., cb_context=2) at /home/nckempke/callbacks/COPT.jl/test/MathOptInterface/callback.jl:42
#25 0x00007fffa17459f6 in julia_#55_393 (cb_data=..., cb_context=0) at /home/nckempke/callbacks/COPT.jl/src/MOI/callback.jl:75
#26 0x00007fffa1745a2c in jfptr_#55_394 ()
#27 0x00007ffff717d994 in _jl_invoke (F=0x7fffefd7fd90, args=0x7fff8913f950, nargs=2, mfunc=0x7ffff081f710, world=33494) at /home/nckempke/git/julia/src/gf.c:2739
#28 0x00007ffff717e439 in ijl_apply_generic (F=0x7fffefd7fd90, args=0x7fff8913f950, nargs=2) at /home/nckempke/git/julia/src/gf.c:2940
#29 0x00007fffa1732ef8 in julia__copt_callback_wrapper_88 (p_model=<optimized out>, cb_data=93825015353152, cb_context=2, p_user_data=<optimized out>)
    at /home/nckempke/callbacks/COPT.jl/src/MOI/callback.jl:38
#30 0x00007fffa173d9c9 in jlcapi.copt_callback_wrapper_91 ()
...
COPT frames
...
#42 0x00007ffff7f8d609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#43 0x00007ffff7eb2133 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

The changes in the files are 
1) adding println to `src/MOI/callback.jl` (with flush after) 
2) reducing the actual test to only `test_CallbackFunction_callback_LazyConstraint` in `test/MathOptInterface/callback.jl` to make it easier to reproduce (as this deadlock does not always happen)
3) Set `Threads` to 2 in `callback_simple_model`.
4) I named the files `.jl.txt` as the GUI of github told me `.jl` was not a supported format...

[callback.jl.txt](https://github.com/COPT-Public/COPT.jl/files/11476429/callback.jl.txt)
[callback.jl.txt](https://github.com/COPT-Public/COPT.jl/files/11476442/callback.jl.txt)

Reply:
```
#1  __pthread_cond_wait_common (abstime=0x0, clockid=0, mutex=0x7fff840057c0, cond=0x7fff840057e8) at pthread_cond_wait.c:508
#2  __pthread_cond_wait (cond=0x7fff840057e8, mutex=0x7fff840057c0) at pthread_cond_wait.c:647
#3  0x00007ffff7290938 in uv_cond_wait (cond=0x7fff840057e8, mutex=0x7fff840057c0) at src/unix/thread.c:883
#4  0x00007ffff71e2261 in ijl_task_get_next (trypoptask=0x7fffe3a73db0 <jl_system_image_data+7600880>, q=0x7fff8c6cc050, checkempty=0x7fffe55c38c0 <jl_system_image_data+36238848>)
    at /home/nckempke/git/julia/src/partr.c:421
#5  0x00007fffe2b6f378 in julia_poptask_33132 (W=...) at task.jl:974
```
This part suggests that the Julia callback is waiting for a lock on `stdout`, so I assume there's some race condition where the thread from C enters Julia before the last one has released the lock, and then then things get corrupted. 

Normally printing happens in a separate task but perhaps it's different when coming from C. Does it block if you try `@async println("abc")`?

Since this is a new feature and pretty niche (printing in callbacks from a secondary thread started in C), it may also just be a plain bug. 

Most people will probably add prints to their callbacks during the initial development, so perhaps it's best to also set threads to 1 in Julia 1.9 when setting `COPT.CallbackFunction`, and document that users can set a larger number of threads, if they promise not to use printing?

Reply:
It is still blocking sometimes with the same trace.

I agree - users will definitely trigger this if not warned. So yes, that is probably the best solution for now.